### PR TITLE
Feature/btc 295

### PR DESCRIPTION
**Pull request by Mr-Leshiy**

### CODE REVIEW

File: include/veriblock/popmanager.hpp [hunk 1]
```diff
@@ -50,19 +51,22 @@ struct PopManager {
    * @throws may throw if out of memory. In this case, payloads also will be
    * reverted.
    */
-  bool addPayloads(const Payloads& payloads, ValidationState& state);
+  bool addPayloads(const Payloads& payloads,
+                   std::shared_ptr<StateChange> stateChange,
```
Comment (Warchant) [inline include/veriblock/popmanager.hpp]: pass by ref

File: include/veriblock/popmanager.hpp [hunk 2]
```diff
@@ -50,19 +51,22 @@ struct PopManager {
...
    */
-  void removePayloads(const Payloads& payloads) noexcept;
+  void removePayloads(const Payloads& payloads,
+                      std::shared_ptr<StateChange> stateChange) noexcept;
 
   bool hasUncommittedChanges() const noexcept;
 
   /// removes all uncommitted payloads
-  void rollback() noexcept;
+  void rollback(std::shared_ptr<StateChange> stateChange) noexcept;
```
Comment (Warchant) [inline include/veriblock/popmanager.hpp]: also here, and below

File: include/veriblock/storage/endorsement_repository_inmem.hpp [hunk 3]
```diff
@@ -148,6 +148,10 @@ struct EndorsementRepositoryInmem : public EndorsementRepository<Endorsement> {
     return ret;
   }
 
+  std::unique_ptr<EndorsementWriteBatch<Endorsement>> newBatch() {
+    return nullptr;
```
Comment (Warchant) [inline include/veriblock/storage/endorsement_repository_inmem.hpp]: ?

File: include/veriblock/storage/endorsement_repository_rocks.hpp [hunk 4]
```diff
@@ -64,239 +64,156 @@ struct EndorsementCursorRocks
...
   }
 
  private:
   std::shared_ptr<rocksdb::DB> _db;
-  // We have two maps [endorsed_block_hash] -> [endorsed_id array]
-  // endorsed_id array is stored using the raw representation described below :
-  // number_of_endorsment_ids(uint32_t), endorsment_id_1_raw_bytes,
-  // endorsment_id_2_raw_bytes, ...
-  // And the another one map is [endorsed_id] -> [endorsement]
+  // [endorsed_id] -> [endorsement]
```
Comment (Warchant) [inline include/veriblock/storage/endorsement_repository_rocks.hpp]:
```suggestion
  // [endorsement id] -> [endorsement]
```

File: src/popmanager.cpp [hunk 5]
```diff
@@ -5,110 +5,139 @@
...
   }
 
   // secondly, add VBK endorsements in BTC
   btce_->put(vtb);
+  stateChange->putBtcEndorsement(vtb);
 
   // thirdly, add vbk context blocks
   for (const auto& block : vtb.context) {
+    BlockIndex<VbkBlock> index;
     if (!vbk_->acceptBlock(block, state)) {
```
Comment (Warchant) [inline src/popmanager.cpp]: maybe this?
```suggestion
    if (!vbk_->acceptBlock(block, state, &index)) {
```

File: include/veriblock/storage/endorsement_repository_inmem.hpp (hunk 3 above)
Comment (Mr-Leshiy) [reply include/veriblock/storage/endorsement_repository_inmem.hpp]: We dont have inmem WriteBatch realisation, and generally dont use it, so here it is a mock of this method.

File: include/veriblock/storage/endorsement_repository_rocks.hpp (hunk 4 above)
Comment (Mr-Leshiy) [reply include/veriblock/storage/endorsement_repository_rocks.hpp]: fixed

**Pull request merged**